draft-ietf-jsonpath-base.md

```diff
@@ -1411,82 +1410,101 @@ a type system is first introduced.
 {: #tbl-typerep title="Concrete representations of abstract instances"}
 
-The following subtype relationships depend on coercion:
+### Type Conversions {#type-conv}
+
+The following type conversions may occur:
 
-* `OptionalNodeType` is a subtype of `OptionalValueType` via coercion since the `OptionalNodeType` instance `Node(n)` can be coerced to
-the `OptionalValueType` instance `Value(v)`, where `v` is the value of the node `n`.
-* `OptionalNodeType` is a subtype of `OptionalNodesType` via coercion since the `OptionalNodeType` instance `Node(n)` can be coerced to
-the `OptionalNodesType` instance `Nodes(l)`, where `l` is a nodelist consisting of just the node `n`.
+* A member of `NodelistType` may be converted to a `ValueType` as follows:
+  * If the nodelist contains a single node, the conversion result is the value of the node.
+  * If the nodelist is empty or contains multiple nodes, the conversion result is `Nothing`.
+* A member of `NodelistType` may be converted to a `BooleanType` as follows:
+  * If the nodelist contains one or more nodes, the conversion result is `BooleanTrue`.
+  * If the nodelist is empty, the conversion result is `BooleanFalse`.
 
 The type correctness of function expressions can now be defined in terms of this type system.
 
 ### Type Correctness of Function Expressions
 
-A function expression is correctly typed if all the following are true:
-
-* If it occurs directly in a test expression, the function
-is defined to have a result type in `ST(OptionalNodesType)`.
-or to have a result type in `ST(OptionalBooleanType)`.
-* If it occurs as a `comparable` in a comparison, the function
-is defined to have a result type in `ST(OptionalNodeOrValueType)`.
-* For it and any function expression it contains,
-each argument of the function matches the defined type of the argument
-according to one of the following rules:
-  * The argument is a function expression with defined result type
-    that is the same as, or a subtype of, the defined type of the argument.
-  * The argument is a literal primitive value and the defined type of the  argument is `Value` or any type of which `Value` is a subtype.
-  * The argument is a Singular Path and the defined type of the argument is `OptionalNodeType` or any type of which `OptionalNodeType` is a subtype.
-  * The argument is a `filter-path` (which includes Singular Paths) and the defined type of the argument is `OptionalNodesType`.
+A function expression is typed based on its return, and may only appear in expressions as follows:
+
+* If the function return is of type `ValueType`, it may only
+  appear as a `comparable` in a comparison.
+* If the function return is of type `BooleanType`, it may
+  only appear in a `test-expr`.
+* If the function return is of type `NodelistType`, it may
+  appear either as a `comparable` in a comparison or in
+  a `test-expr` as members of `NodelistType` can be converted
+  to either `ValueType` or `BooleanType`.
```

Comment:
I'd prefer `NodelistType` to be usable in comparisons and test expressions without conversion.

Reply:
Please explain how this can be?  Comparisons have values as operands and output booleans.  Test expressions have booleans has operands and output booleans.  Without conversion, nodelists would not be valid.